File: rsa.js
const crypto = require('crypto');

function generateKeys() {
    const { publicKey, privateKey } = crypto.generateKeyPairSync('rsa', {
        modulusLength: 2048,
        publicKeyEncoding: { type: 'spki', format: 'der' },
        privateKeyEncoding: { type: 'pkcs8', format: 'der' }
    });

    return { 
        publicKey: publicKey.toString("base64").replace(/\r?\n|\r/g, ''), 
        privateKey: privateKey.toString("base64").replace(/\r?\n|\r/g, '') 
    };
}

function encrypt(publicKeyStr, value) {
    const publicKey = crypto.createPublicKey({
        key: Buffer.from(publicKeyStr, 'base64'),
        padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
        format: 'der',
        type:'spki'
    });
    
    const encryptedBuffer = crypto.publicEncrypt({
        key: publicKey,
        padding: crypto.constants.RSA_PKCS1_OAEP_PADDING
    }, Buffer.from(value)); 

    return encryptedBuffer.toString('base64');
}

function decrypt(privateKeyStr, value) {
    const publicKeyFromStr = crypto.createPrivateKey({
        key: Buffer.from(privateKeyStr, 'base64'),
        padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
        format: 'der',
        type:'pkcs8'
    });

    const decryptedBuffer = crypto.privateDecrypt({
        key: publicKeyFromStr,
        padding: crypto.constants.RSA_PKCS1_OAEP_PADDING  
    }, Buffer.from(value, 'base64'));

    return decryptedBuffer.toString('utf8');
}

function main() {
    const originalMessage = "This `s RSA Encryption Using PUBL1C/PRIVAT3 Key$";
    console.log("Original Message: ", originalMessage);
    
    const keys = generateKeys();
    console.log("Public Key: ", keys.publicKey);
    console.log("Private Key: ", keys.privateKey);
    
    const encryptedData = encrypt(keys.publicKey, originalMessage);
    console.log("Encrypted Data: ", encryptedData);
 
    const decryptedData = decrypt(keys.privateKey, encryptedData);
    console.log("Decrypted Data: ", decryptedData);
}

main();

File: RSA.java
import java.nio.charset.StandardCharsets;
import java.security.*;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;
import javax.crypto.Cipher;

public class RSA {

    public static String[] generateKeys() {
    	try {
            KeyPairGenerator keyGen = KeyPairGenerator.getInstance("RSA");
            keyGen.initialize(2048);
            
            KeyPair keys = keyGen.generateKeyPair();
            
            return new String[] {
        		Base64.getEncoder().encodeToString(keys.getPublic().getEncoded()),
        		Base64.getEncoder().encodeToString(keys.getPrivate().getEncoded()),
            };    		
    	} catch(NoSuchAlgorithmException e) {
    		return null;
    	}
    }
  
    public static String encrypt(String publicKeyStr, String value) {
    	try {
        	byte[] publicKeyBytes = Base64.getDecoder().decode(publicKeyStr.getBytes());
        	byte[] valueBytes = value.getBytes(StandardCharsets.UTF_8);
        	
            X509EncodedKeySpec keySpec = new X509EncodedKeySpec(publicKeyBytes);
            KeyFactory keyFactory = KeyFactory.getInstance("RSA");
            PublicKey publicKey = keyFactory.generatePublic(keySpec);

            Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-1AndMGF1Padding");
            cipher.init(Cipher.ENCRYPT_MODE, publicKey);

            return Base64.getEncoder().encodeToString(cipher.doFinal(valueBytes));    		
    	} catch(Exception e) {
    		return null;
    	}
    }
    
    
    public static String decrypt(String privateKeyStr, String value) {
    	try {
        	byte[] privateKeyBytes = Base64.getDecoder().decode(privateKeyStr.getBytes());
        	byte[] valueBytes = Base64.getDecoder().decode(value.getBytes());
            
    		PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(privateKeyBytes);
            KeyFactory keyFactory = KeyFactory.getInstance("RSA");
            PrivateKey privateKey = keyFactory.generatePrivate(keySpec);

            Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-1AndMGF1Padding");
            cipher.init(Cipher.DECRYPT_MODE, privateKey);

            return new String(cipher.doFinal(valueBytes), StandardCharsets.UTF_8);    		
    	} catch(Exception e) {
    		return null;
    	}
    }
    
    public static void main(String[] args) {
        String originalMessage = "This `s RSA Encryption Using PUBL1C/PRIVAT3 Key$";
        System.out.println("Original Message: " + originalMessage);
        
        String[] keys = generateKeys();
        
        if(keys == null) {
        	System.out.println("Unable to generate keys.");
        	return;
        }
        
        String publicKey = keys[0];
        String privateKey = keys[1];
        System.out.println("Public Key: " + publicKey);
        System.out.println("Private Key: " + privateKey);
        
        String encryptedData = encrypt(publicKey, originalMessage);
        System.out.println("Encrypted Data: " + encryptedData);

        String decryptedData = decrypt(privateKey, encryptedData);
        System.out.println("Decrypted Data: " + decryptedData);
    }
}
